{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jho5ze/.local/lib/python3.8/site-packages/pysradb/utils.py:14: TqdmExperimentalWarning: Using `tqdm.autonotebook.tqdm` in notebook mode. Use `tqdm.tqdm` instead to force console mode (e.g. in jupyter console)\n",
      "  from tqdm.autonotebook import tqdm\n"
     ]
    }
   ],
   "source": [
    "import sqlite3\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "# import pysradb as psra\n",
    "from pysradb.search import SraSearch, EnaSearch, GeoSearch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1000/1000 [00:34<00:00, 29.37it/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>experiment_title</th>\n",
       "      <th>sample_taxon_id</th>\n",
       "      <th>sample_scientific_name</th>\n",
       "      <th>experiment_library_strategy</th>\n",
       "      <th>experiment_library_source</th>\n",
       "      <th>experiment_library_selection</th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>sample_alias</th>\n",
       "      <th>experiment_instrument_model</th>\n",
       "      <th>pool_member_spots</th>\n",
       "      <th>run_1_size</th>\n",
       "      <th>run_1_accession</th>\n",
       "      <th>run_1_total_spots</th>\n",
       "      <th>run_1_total_bases</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRP344499</td>\n",
       "      <td>SRX12967959</td>\n",
       "      <td>CDSs of Adiantum flabellulatum gametocyte</td>\n",
       "      <td>872337</td>\n",
       "      <td>Adiantum flabellulatum</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>Oligo-dT</td>\n",
       "      <td>SRS10909271</td>\n",
       "      <td>MERGE ALL SAMPLE</td>\n",
       "      <td>Sequel</td>\n",
       "      <td>231705</td>\n",
       "      <td>52779868</td>\n",
       "      <td>SRR16768024</td>\n",
       "      <td>231705</td>\n",
       "      <td>194628012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRP344426</td>\n",
       "      <td>SRX12963949</td>\n",
       "      <td>GSM5668884: GRA6_TEFF_4; Mus musculus; RNA-Seq</td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>SRS10905616</td>\n",
       "      <td>GSM5668884</td>\n",
       "      <td>Illumina HiSeq 4000</td>\n",
       "      <td>31168024</td>\n",
       "      <td>589160072</td>\n",
       "      <td>SRR16763848</td>\n",
       "      <td>31168024</td>\n",
       "      <td>1527233176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRP344426</td>\n",
       "      <td>SRX12963948</td>\n",
       "      <td>GSM5668883: GRA6_TEFF_3; Mus musculus; RNA-Seq</td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>SRS10905615</td>\n",
       "      <td>GSM5668883</td>\n",
       "      <td>Illumina HiSeq 4000</td>\n",
       "      <td>26356295</td>\n",
       "      <td>752236799</td>\n",
       "      <td>SRR16763849</td>\n",
       "      <td>26356295</td>\n",
       "      <td>1291458455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRP344426</td>\n",
       "      <td>SRX12963947</td>\n",
       "      <td>GSM5668882: GRA6_TEFF_2; Mus musculus; RNA-Seq</td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>SRS10905614</td>\n",
       "      <td>GSM5668882</td>\n",
       "      <td>Illumina HiSeq 4000</td>\n",
       "      <td>24834727</td>\n",
       "      <td>709243482</td>\n",
       "      <td>SRR16763850</td>\n",
       "      <td>24834727</td>\n",
       "      <td>1216901623</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRP344426</td>\n",
       "      <td>SRX12963946</td>\n",
       "      <td>GSM5668881: GRA6_TEFF_1; Mus musculus; RNA-Seq</td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>SRS10905612</td>\n",
       "      <td>GSM5668881</td>\n",
       "      <td>Illumina HiSeq 4000</td>\n",
       "      <td>25106867</td>\n",
       "      <td>718367618</td>\n",
       "      <td>SRR16763851</td>\n",
       "      <td>25106867</td>\n",
       "      <td>1230236483</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>995</th>\n",
       "      <td>SRP344246</td>\n",
       "      <td>SRX12902234</td>\n",
       "      <td>GSM5666002: MCF7 cells treated with 0.96 uM SB...</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>SRS10844446</td>\n",
       "      <td>GSM5666002</td>\n",
       "      <td>NextSeq 550</td>\n",
       "      <td>1018782</td>\n",
       "      <td>62318981</td>\n",
       "      <td>SRR16701685</td>\n",
       "      <td>1018782</td>\n",
       "      <td>50939100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>996</th>\n",
       "      <td>SRP344246</td>\n",
       "      <td>SRX12902233</td>\n",
       "      <td>GSM5666001: MCF7 cells treated with 24.96 uM M...</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>SRS10844441</td>\n",
       "      <td>GSM5666001</td>\n",
       "      <td>NextSeq 550</td>\n",
       "      <td>788372</td>\n",
       "      <td>48303415</td>\n",
       "      <td>SRR16701686</td>\n",
       "      <td>788372</td>\n",
       "      <td>39418600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>997</th>\n",
       "      <td>SRP344246</td>\n",
       "      <td>SRX12902232</td>\n",
       "      <td>GSM5666000: MCF7 cells treated with 10 uM SB1-...</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>SRS10844438</td>\n",
       "      <td>GSM5666000</td>\n",
       "      <td>NextSeq 550</td>\n",
       "      <td>452136</td>\n",
       "      <td>27744646</td>\n",
       "      <td>SRR16701687</td>\n",
       "      <td>452136</td>\n",
       "      <td>22606800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>998</th>\n",
       "      <td>SRP344246</td>\n",
       "      <td>SRX12902231</td>\n",
       "      <td>GSM5666567: MDAMB231 cells with knockout of NT...</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>SRS10844437</td>\n",
       "      <td>GSM5666567</td>\n",
       "      <td>NextSeq 550</td>\n",
       "      <td>790208</td>\n",
       "      <td>54898802</td>\n",
       "      <td>SRR16701688</td>\n",
       "      <td>790208</td>\n",
       "      <td>51363520</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999</th>\n",
       "      <td>SRP344246</td>\n",
       "      <td>SRX12902230</td>\n",
       "      <td>GSM5666566: MDAMB231 cells with knockout of MI...</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>SRS10844439</td>\n",
       "      <td>GSM5666566</td>\n",
       "      <td>NextSeq 550</td>\n",
       "      <td>393189</td>\n",
       "      <td>27252057</td>\n",
       "      <td>SRR16701689</td>\n",
       "      <td>393189</td>\n",
       "      <td>25557285</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1000 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    study_accession experiment_accession  \\\n",
       "0         SRP344499          SRX12967959   \n",
       "1         SRP344426          SRX12963949   \n",
       "2         SRP344426          SRX12963948   \n",
       "3         SRP344426          SRX12963947   \n",
       "4         SRP344426          SRX12963946   \n",
       "..              ...                  ...   \n",
       "995       SRP344246          SRX12902234   \n",
       "996       SRP344246          SRX12902233   \n",
       "997       SRP344246          SRX12902232   \n",
       "998       SRP344246          SRX12902231   \n",
       "999       SRP344246          SRX12902230   \n",
       "\n",
       "                                      experiment_title sample_taxon_id  \\\n",
       "0            CDSs of Adiantum flabellulatum gametocyte          872337   \n",
       "1       GSM5668884: GRA6_TEFF_4; Mus musculus; RNA-Seq           10090   \n",
       "2       GSM5668883: GRA6_TEFF_3; Mus musculus; RNA-Seq           10090   \n",
       "3       GSM5668882: GRA6_TEFF_2; Mus musculus; RNA-Seq           10090   \n",
       "4       GSM5668881: GRA6_TEFF_1; Mus musculus; RNA-Seq           10090   \n",
       "..                                                 ...             ...   \n",
       "995  GSM5666002: MCF7 cells treated with 0.96 uM SB...            9606   \n",
       "996  GSM5666001: MCF7 cells treated with 24.96 uM M...            9606   \n",
       "997  GSM5666000: MCF7 cells treated with 10 uM SB1-...            9606   \n",
       "998  GSM5666567: MDAMB231 cells with knockout of NT...            9606   \n",
       "999  GSM5666566: MDAMB231 cells with knockout of MI...            9606   \n",
       "\n",
       "     sample_scientific_name experiment_library_strategy  \\\n",
       "0    Adiantum flabellulatum                     RNA-Seq   \n",
       "1              Mus musculus                     RNA-Seq   \n",
       "2              Mus musculus                     RNA-Seq   \n",
       "3              Mus musculus                     RNA-Seq   \n",
       "4              Mus musculus                     RNA-Seq   \n",
       "..                      ...                         ...   \n",
       "995            Homo sapiens                     RNA-Seq   \n",
       "996            Homo sapiens                     RNA-Seq   \n",
       "997            Homo sapiens                     RNA-Seq   \n",
       "998            Homo sapiens                     RNA-Seq   \n",
       "999            Homo sapiens                     RNA-Seq   \n",
       "\n",
       "    experiment_library_source experiment_library_selection sample_accession  \\\n",
       "0              TRANSCRIPTOMIC                     Oligo-dT      SRS10909271   \n",
       "1              TRANSCRIPTOMIC                         cDNA      SRS10905616   \n",
       "2              TRANSCRIPTOMIC                         cDNA      SRS10905615   \n",
       "3              TRANSCRIPTOMIC                         cDNA      SRS10905614   \n",
       "4              TRANSCRIPTOMIC                         cDNA      SRS10905612   \n",
       "..                        ...                          ...              ...   \n",
       "995            TRANSCRIPTOMIC                         cDNA      SRS10844446   \n",
       "996            TRANSCRIPTOMIC                         cDNA      SRS10844441   \n",
       "997            TRANSCRIPTOMIC                         cDNA      SRS10844438   \n",
       "998            TRANSCRIPTOMIC                         cDNA      SRS10844437   \n",
       "999            TRANSCRIPTOMIC                         cDNA      SRS10844439   \n",
       "\n",
       "         sample_alias experiment_instrument_model pool_member_spots  \\\n",
       "0    MERGE ALL SAMPLE                      Sequel            231705   \n",
       "1          GSM5668884         Illumina HiSeq 4000          31168024   \n",
       "2          GSM5668883         Illumina HiSeq 4000          26356295   \n",
       "3          GSM5668882         Illumina HiSeq 4000          24834727   \n",
       "4          GSM5668881         Illumina HiSeq 4000          25106867   \n",
       "..                ...                         ...               ...   \n",
       "995        GSM5666002                 NextSeq 550           1018782   \n",
       "996        GSM5666001                 NextSeq 550            788372   \n",
       "997        GSM5666000                 NextSeq 550            452136   \n",
       "998        GSM5666567                 NextSeq 550            790208   \n",
       "999        GSM5666566                 NextSeq 550            393189   \n",
       "\n",
       "    run_1_size run_1_accession run_1_total_spots run_1_total_bases  \n",
       "0     52779868     SRR16768024            231705         194628012  \n",
       "1    589160072     SRR16763848          31168024        1527233176  \n",
       "2    752236799     SRR16763849          26356295        1291458455  \n",
       "3    709243482     SRR16763850          24834727        1216901623  \n",
       "4    718367618     SRR16763851          25106867        1230236483  \n",
       "..         ...             ...               ...               ...  \n",
       "995   62318981     SRR16701685           1018782          50939100  \n",
       "996   48303415     SRR16701686            788372          39418600  \n",
       "997   27744646     SRR16701687            452136          22606800  \n",
       "998   54898802     SRR16701688            790208          51363520  \n",
       "999   27252057     SRR16701689            393189          25557285  \n",
       "\n",
       "[1000 rows x 16 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x = SraSearch(return_max=1000, query=\"Transcriptome Analysis\")\n",
    "x.search()\n",
    "x = x.get_df()\n",
    "x"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Pysradb approach"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1/1 [00:00<00:00,  5.33it/s]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array(['SRP235283'], dtype=object)"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x = SraSearch(return_max=150, query=\"MG1655 RNA-Seq paired stranded\") #Escherichia dRNA-Seq\")\n",
    "x.search()\n",
    "x = x.get_df()\n",
    "# x.sample_scientific_name = x.sample_scientific_name.apply(lambda row: row.replace(\" str. \", \" strain=\"))#.run_1_accession.unique()\n",
    "x.study_accession.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>experiment_title</th>\n",
       "      <th>sample_taxon_id</th>\n",
       "      <th>sample_scientific_name</th>\n",
       "      <th>experiment_library_strategy</th>\n",
       "      <th>experiment_library_source</th>\n",
       "      <th>experiment_library_selection</th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>sample_alias</th>\n",
       "      <th>experiment_instrument_model</th>\n",
       "      <th>pool_member_spots</th>\n",
       "      <th>run_1_size</th>\n",
       "      <th>run_1_accession</th>\n",
       "      <th>run_1_total_spots</th>\n",
       "      <th>run_1_total_bases</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRP235283</td>\n",
       "      <td>SRX7291196</td>\n",
       "      <td>GSM4211192: WT rep2_21°C; Escherichia coli; RN...</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>SRS5784357</td>\n",
       "      <td>GSM4211192</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>17461337</td>\n",
       "      <td>790322623</td>\n",
       "      <td>SRR10612035</td>\n",
       "      <td>17461337</td>\n",
       "      <td>1729601877</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  study_accession experiment_accession  \\\n",
       "0       SRP235283           SRX7291196   \n",
       "\n",
       "                                    experiment_title sample_taxon_id  \\\n",
       "0  GSM4211192: WT rep2_21°C; Escherichia coli; RN...             562   \n",
       "\n",
       "  sample_scientific_name experiment_library_strategy  \\\n",
       "0       Escherichia coli                     RNA-Seq   \n",
       "\n",
       "  experiment_library_source experiment_library_selection sample_accession  \\\n",
       "0            TRANSCRIPTOMIC                         cDNA       SRS5784357   \n",
       "\n",
       "  sample_alias experiment_instrument_model pool_member_spots run_1_size  \\\n",
       "0   GSM4211192         Illumina HiSeq 2000          17461337  790322623   \n",
       "\n",
       "  run_1_accession run_1_total_spots run_1_total_bases  \n",
       "0     SRR10612035          17461337        1729601877  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "x[x.study_accession == \"SRP235283\"].to_csv(\"Escherichia_coli_representative_sample_info.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "x[(x.run_1_accession.apply(lambda row: row.replace(\"SRR\", \"\")).astype(int) >=794848) & (x.run_1_accession.apply(lambda row: row.replace(\"SRR\", \"\")).astype(int) <= 794850)].to_csv(\"Escherichia_query_info.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "x.head(20).to_csv(\"Helicobacter_query_info.csv\") #[\"experiment_title\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = EnaSearch(2, 20, \"tuberculosis RNA\")\n",
    "x.search()\n",
    "x = x.get_df()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>experiment_title</th>\n",
       "      <th>description</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>library_source</th>\n",
       "      <th>library_selection</th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>sample_title</th>\n",
       "      <th>instrument_model</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>read_count</th>\n",
       "      <th>base_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PRJNA239223</td>\n",
       "      <td>SRX475349</td>\n",
       "      <td>Illumina HiSeq 2000 sequencing; GSM1333720: Ye...</td>\n",
       "      <td>Illumina HiSeq 2000 sequencing; GSM1333720: Ye...</td>\n",
       "      <td>502800</td>\n",
       "      <td>Yersinia pseudotuberculosis YPIII</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>SAMN02650479</td>\n",
       "      <td>Yersinia pseudotuberculosis RNA-seq_early infe...</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>SRR1175455</td>\n",
       "      <td>88333445</td>\n",
       "      <td>17843355890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PRJNA239223</td>\n",
       "      <td>SRX475349</td>\n",
       "      <td>Illumina HiSeq 2000 sequencing; GSM1333720: Ye...</td>\n",
       "      <td>Illumina HiSeq 2000 sequencing; GSM1333720: Ye...</td>\n",
       "      <td>502800</td>\n",
       "      <td>Yersinia pseudotuberculosis YPIII</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>SAMN02650479</td>\n",
       "      <td>Yersinia pseudotuberculosis RNA-seq_early infe...</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>SRR1175456</td>\n",
       "      <td>70676677</td>\n",
       "      <td>14276688754</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PRJNA239223</td>\n",
       "      <td>SRX475350</td>\n",
       "      <td>Illumina HiSeq 2000 sequencing; GSM1333721: Y....</td>\n",
       "      <td>Illumina HiSeq 2000 sequencing; GSM1333721: Y....</td>\n",
       "      <td>502800</td>\n",
       "      <td>Yersinia pseudotuberculosis YPIII</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>SAMN02650477</td>\n",
       "      <td>Y. pseudotuberculosis RNA-seq _persistent infe...</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>SRR1175457</td>\n",
       "      <td>96794171</td>\n",
       "      <td>19552422542</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PRJNA239223</td>\n",
       "      <td>SRX475350</td>\n",
       "      <td>Illumina HiSeq 2000 sequencing; GSM1333721: Y....</td>\n",
       "      <td>Illumina HiSeq 2000 sequencing; GSM1333721: Y....</td>\n",
       "      <td>502800</td>\n",
       "      <td>Yersinia pseudotuberculosis YPIII</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>SAMN02650477</td>\n",
       "      <td>Y. pseudotuberculosis RNA-seq _persistent infe...</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>SRR1175458</td>\n",
       "      <td>97439250</td>\n",
       "      <td>19682728500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PRJNA239223</td>\n",
       "      <td>SRX475351</td>\n",
       "      <td>Illumina HiSeq 2000 sequencing; GSM1333722: Y....</td>\n",
       "      <td>Illumina HiSeq 2000 sequencing; GSM1333722: Y....</td>\n",
       "      <td>502800</td>\n",
       "      <td>Yersinia pseudotuberculosis YPIII</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>SAMN02650480</td>\n",
       "      <td>Y. pseudotuberculosis RNA-seq_early infection,...</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>SRR1175459</td>\n",
       "      <td>87517953</td>\n",
       "      <td>17678626506</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  study_accession experiment_accession  \\\n",
       "0     PRJNA239223            SRX475349   \n",
       "1     PRJNA239223            SRX475349   \n",
       "2     PRJNA239223            SRX475350   \n",
       "3     PRJNA239223            SRX475350   \n",
       "4     PRJNA239223            SRX475351   \n",
       "\n",
       "                                    experiment_title  \\\n",
       "0  Illumina HiSeq 2000 sequencing; GSM1333720: Ye...   \n",
       "1  Illumina HiSeq 2000 sequencing; GSM1333720: Ye...   \n",
       "2  Illumina HiSeq 2000 sequencing; GSM1333721: Y....   \n",
       "3  Illumina HiSeq 2000 sequencing; GSM1333721: Y....   \n",
       "4  Illumina HiSeq 2000 sequencing; GSM1333722: Y....   \n",
       "\n",
       "                                         description  tax_id  \\\n",
       "0  Illumina HiSeq 2000 sequencing; GSM1333720: Ye...  502800   \n",
       "1  Illumina HiSeq 2000 sequencing; GSM1333720: Ye...  502800   \n",
       "2  Illumina HiSeq 2000 sequencing; GSM1333721: Y....  502800   \n",
       "3  Illumina HiSeq 2000 sequencing; GSM1333721: Y....  502800   \n",
       "4  Illumina HiSeq 2000 sequencing; GSM1333722: Y....  502800   \n",
       "\n",
       "                     scientific_name library_strategy  library_source  \\\n",
       "0  Yersinia pseudotuberculosis YPIII          RNA-Seq  TRANSCRIPTOMIC   \n",
       "1  Yersinia pseudotuberculosis YPIII          RNA-Seq  TRANSCRIPTOMIC   \n",
       "2  Yersinia pseudotuberculosis YPIII          RNA-Seq  TRANSCRIPTOMIC   \n",
       "3  Yersinia pseudotuberculosis YPIII          RNA-Seq  TRANSCRIPTOMIC   \n",
       "4  Yersinia pseudotuberculosis YPIII          RNA-Seq  TRANSCRIPTOMIC   \n",
       "\n",
       "  library_selection sample_accession  \\\n",
       "0              cDNA     SAMN02650479   \n",
       "1              cDNA     SAMN02650479   \n",
       "2              cDNA     SAMN02650477   \n",
       "3              cDNA     SAMN02650477   \n",
       "4              cDNA     SAMN02650480   \n",
       "\n",
       "                                        sample_title     instrument_model  \\\n",
       "0  Yersinia pseudotuberculosis RNA-seq_early infe...  Illumina HiSeq 2000   \n",
       "1  Yersinia pseudotuberculosis RNA-seq_early infe...  Illumina HiSeq 2000   \n",
       "2  Y. pseudotuberculosis RNA-seq _persistent infe...  Illumina HiSeq 2000   \n",
       "3  Y. pseudotuberculosis RNA-seq _persistent infe...  Illumina HiSeq 2000   \n",
       "4  Y. pseudotuberculosis RNA-seq_early infection,...  Illumina HiSeq 2000   \n",
       "\n",
       "  run_accession read_count   base_count  \n",
       "0    SRR1175455   88333445  17843355890  \n",
       "1    SRR1175456   70676677  14276688754  \n",
       "2    SRR1175457   96794171  19552422542  \n",
       "3    SRR1175458   97439250  19682728500  \n",
       "4    SRR1175459   87517953  17678626506  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Sqlite approach"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[('metaInfo',), ('submission',), ('study',), ('sample',), ('experiment',), ('run',), ('sra',), ('sra_ft',), ('sra_ft_content',), ('sra_ft_segments',), ('sra_ft_segdir',), ('col_desc',), ('fastq',)]\n"
     ]
    }
   ],
   "source": [
    "con = sqlite3.connect(\"/sfs/qumulo/qproject/biocomplexity/nssac/operons/SRAmetadb.sqlite\")\n",
    "cursor = con.cursor()\n",
    "cursor.execute(\"select name from sqlite_master where type='table'\")\n",
    "print(cursor.fetchall())\n",
    "# cur = con.cursor()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['sample_ID', 'sample_alias', 'sample_accession', 'broker_name',\n",
       "       'center_name', 'taxon_id', 'scientific_name', 'common_name',\n",
       "       'anonymized_name', 'individual_name', 'description', 'sra_link',\n",
       "       'sample_url_link', 'xref_link', 'sample_entrez_link', 'ddbj_link',\n",
       "       'ena_link', 'sample_attribute', 'submission_accession',\n",
       "       'sradb_updated'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_sql_query(\"SELECT * FROM sample\", con)\n",
    "df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_ID</th>\n",
       "      <th>sample_alias</th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>broker_name</th>\n",
       "      <th>center_name</th>\n",
       "      <th>taxon_id</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>common_name</th>\n",
       "      <th>anonymized_name</th>\n",
       "      <th>individual_name</th>\n",
       "      <th>description</th>\n",
       "      <th>sra_link</th>\n",
       "      <th>sample_url_link</th>\n",
       "      <th>xref_link</th>\n",
       "      <th>sample_entrez_link</th>\n",
       "      <th>ddbj_link</th>\n",
       "      <th>ena_link</th>\n",
       "      <th>sample_attribute</th>\n",
       "      <th>submission_accession</th>\n",
       "      <th>sradb_updated</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.0</td>\n",
       "      <td>SAMD00016353</td>\n",
       "      <td>DRS000001</td>\n",
       "      <td>None</td>\n",
       "      <td>BioSample</td>\n",
       "      <td>645657.0</td>\n",
       "      <td>Bacillus subtilis subsp. natto BEST195</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>sample_name: DRS000001 || strain: BEST195 || s...</td>\n",
       "      <td>DRA000001</td>\n",
       "      <td>2016-11-20 05:18:07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2.0</td>\n",
       "      <td>SAMD00002791</td>\n",
       "      <td>DRS000002</td>\n",
       "      <td>None</td>\n",
       "      <td>BioSample</td>\n",
       "      <td>224308.0</td>\n",
       "      <td>Bacillus subtilis subsp. subtilis str. 168</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>sample_name: DRS000002 || strain: str. 168 || ...</td>\n",
       "      <td>DRA000002</td>\n",
       "      <td>2016-11-20 05:18:07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3.0</td>\n",
       "      <td>SAMD00013986</td>\n",
       "      <td>DRS000003</td>\n",
       "      <td>None</td>\n",
       "      <td>BioSample</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>sample_name: DRS000003</td>\n",
       "      <td>DRA000003</td>\n",
       "      <td>2016-11-20 05:18:07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4.0</td>\n",
       "      <td>SAMD00002711</td>\n",
       "      <td>DRS000004</td>\n",
       "      <td>None</td>\n",
       "      <td>BioSample</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>sample_name: DRS000004</td>\n",
       "      <td>DRA000004</td>\n",
       "      <td>2016-11-20 05:18:07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5.0</td>\n",
       "      <td>SAMD00010993</td>\n",
       "      <td>DRS000005</td>\n",
       "      <td>None</td>\n",
       "      <td>BioSample</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>sample_name: DRS000005</td>\n",
       "      <td>DRA000005</td>\n",
       "      <td>2016-11-20 05:18:07</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   sample_ID  sample_alias sample_accession broker_name center_name  taxon_id  \\\n",
       "0        1.0  SAMD00016353        DRS000001        None   BioSample  645657.0   \n",
       "1        2.0  SAMD00002791        DRS000002        None   BioSample  224308.0   \n",
       "2        3.0  SAMD00013986        DRS000003        None   BioSample    9606.0   \n",
       "3        4.0  SAMD00002711        DRS000004        None   BioSample    9606.0   \n",
       "4        5.0  SAMD00010993        DRS000005        None   BioSample    9606.0   \n",
       "\n",
       "                              scientific_name common_name anonymized_name  \\\n",
       "0      Bacillus subtilis subsp. natto BEST195        None            None   \n",
       "1  Bacillus subtilis subsp. subtilis str. 168        None            None   \n",
       "2                                Homo sapiens        None            None   \n",
       "3                                Homo sapiens        None            None   \n",
       "4                                Homo sapiens        None            None   \n",
       "\n",
       "  individual_name description sra_link sample_url_link xref_link  \\\n",
       "0            None        None     None            None      None   \n",
       "1            None        None     None            None      None   \n",
       "2            None        None     None            None      None   \n",
       "3            None        None     None            None      None   \n",
       "4            None        None     None            None      None   \n",
       "\n",
       "  sample_entrez_link ddbj_link ena_link  \\\n",
       "0               None      None     None   \n",
       "1               None      None     None   \n",
       "2               None      None     None   \n",
       "3               None      None     None   \n",
       "4               None      None     None   \n",
       "\n",
       "                                    sample_attribute submission_accession  \\\n",
       "0  sample_name: DRS000001 || strain: BEST195 || s...            DRA000001   \n",
       "1  sample_name: DRS000002 || strain: str. 168 || ...            DRA000002   \n",
       "2                             sample_name: DRS000003            DRA000003   \n",
       "3                             sample_name: DRS000004            DRA000004   \n",
       "4                             sample_name: DRS000005            DRA000005   \n",
       "\n",
       "         sradb_updated  \n",
       "0  2016-11-20 05:18:07  \n",
       "1  2016-11-20 05:18:07  \n",
       "2  2016-11-20 05:18:07  \n",
       "3  2016-11-20 05:18:07  \n",
       "4  2016-11-20 05:18:07  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['study_ID', 'study_alias', 'study_accession', 'study_title',\n",
      "       'study_type', 'study_abstract', 'broker_name', 'center_name',\n",
      "       'center_project_name', 'study_description', 'related_studies',\n",
      "       'primary_study', 'sra_link', 'study_url_link', 'xref_link',\n",
      "       'study_entrez_link', 'ddbj_link', 'ena_link', 'study_attribute',\n",
      "       'submission_accession', 'sradb_updated'],\n",
      "      dtype='object')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_ID</th>\n",
       "      <th>study_alias</th>\n",
       "      <th>study_accession</th>\n",
       "      <th>study_title</th>\n",
       "      <th>study_type</th>\n",
       "      <th>study_abstract</th>\n",
       "      <th>broker_name</th>\n",
       "      <th>center_name</th>\n",
       "      <th>center_project_name</th>\n",
       "      <th>study_description</th>\n",
       "      <th>...</th>\n",
       "      <th>primary_study</th>\n",
       "      <th>sra_link</th>\n",
       "      <th>study_url_link</th>\n",
       "      <th>xref_link</th>\n",
       "      <th>study_entrez_link</th>\n",
       "      <th>ddbj_link</th>\n",
       "      <th>ena_link</th>\n",
       "      <th>study_attribute</th>\n",
       "      <th>submission_accession</th>\n",
       "      <th>sradb_updated</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.0</td>\n",
       "      <td>DRP000001</td>\n",
       "      <td>DRP000001</td>\n",
       "      <td>Bacillus subtilis subsp. natto BEST195 genome ...</td>\n",
       "      <td>Whole Genome Sequencing</td>\n",
       "      <td>&lt;b&gt;&lt;i&gt;Bacillus subtilis&lt;/i&gt; subsp. &lt;i&gt;natto&lt;/i...</td>\n",
       "      <td>None</td>\n",
       "      <td>KEIO</td>\n",
       "      <td>Bacillus subtilis subsp. natto BEST195</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>pubmed: 20398357 || pubmed: 25329997</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>DRA000001</td>\n",
       "      <td>2016-11-20 16:42:15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2.0</td>\n",
       "      <td>DRP000002</td>\n",
       "      <td>DRP000002</td>\n",
       "      <td>Model organism for prokaryotic cell differenti...</td>\n",
       "      <td>Whole Genome Sequencing</td>\n",
       "      <td>Whole genome resequencing of B. subtilis subti...</td>\n",
       "      <td>None</td>\n",
       "      <td>KEIO</td>\n",
       "      <td>Bacillus subtilis subsp. subtilis str. 168</td>\n",
       "      <td>Whole genome resequencing of B. subtilis subti...</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>pubmed: 20398357</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>DRA000002</td>\n",
       "      <td>2017-09-21 20:23:35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3.0</td>\n",
       "      <td>DRP000003</td>\n",
       "      <td>DRP000003</td>\n",
       "      <td>Comprehensive identification and characterizat...</td>\n",
       "      <td>Transcriptome Analysis</td>\n",
       "      <td>Comprehensive identification and characterizat...</td>\n",
       "      <td>None</td>\n",
       "      <td>UT-MGS</td>\n",
       "      <td>Integrative Transcriptome Analysis</td>\n",
       "      <td>Although recent studies have revealed that the...</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>DBTSS: http://dbtss.hgc.jp/</td>\n",
       "      <td>pubmed: 20400770</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>DRA000003</td>\n",
       "      <td>2016-11-20 16:42:16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4.0</td>\n",
       "      <td>DRP000004</td>\n",
       "      <td>DRP000004</td>\n",
       "      <td>Comprehensive identification and characterizat...</td>\n",
       "      <td>Transcriptome Analysis</td>\n",
       "      <td>Comprehensive identification and characterizat...</td>\n",
       "      <td>None</td>\n",
       "      <td>UT-MGS</td>\n",
       "      <td>Integrateve Transcriptome Analysis</td>\n",
       "      <td>Although recent studies have revealed that the...</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>DBTSS: http://dbtss.hgc.jp/</td>\n",
       "      <td>pubmed: 20400770</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>DRA000004</td>\n",
       "      <td>2016-11-20 16:42:16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5.0</td>\n",
       "      <td>DRP000005</td>\n",
       "      <td>DRP000005</td>\n",
       "      <td>Comprehensive identification and characterizat...</td>\n",
       "      <td>Transcriptome Analysis</td>\n",
       "      <td>Comprehensive identification and characterizat...</td>\n",
       "      <td>None</td>\n",
       "      <td>UT-MGS</td>\n",
       "      <td>Integrative Transcriptome Analysis</td>\n",
       "      <td>Although recent studies have revealed that the...</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>DBTSS: http://dbtss.hgc.jp/</td>\n",
       "      <td>pubmed: 20400770</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>DRA000005</td>\n",
       "      <td>2016-11-20 16:42:16</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   study_ID study_alias study_accession  \\\n",
       "0       1.0   DRP000001       DRP000001   \n",
       "1       2.0   DRP000002       DRP000002   \n",
       "2       3.0   DRP000003       DRP000003   \n",
       "3       4.0   DRP000004       DRP000004   \n",
       "4       5.0   DRP000005       DRP000005   \n",
       "\n",
       "                                         study_title               study_type  \\\n",
       "0  Bacillus subtilis subsp. natto BEST195 genome ...  Whole Genome Sequencing   \n",
       "1  Model organism for prokaryotic cell differenti...  Whole Genome Sequencing   \n",
       "2  Comprehensive identification and characterizat...   Transcriptome Analysis   \n",
       "3  Comprehensive identification and characterizat...   Transcriptome Analysis   \n",
       "4  Comprehensive identification and characterizat...   Transcriptome Analysis   \n",
       "\n",
       "                                      study_abstract broker_name center_name  \\\n",
       "0  <b><i>Bacillus subtilis</i> subsp. <i>natto</i...        None        KEIO   \n",
       "1  Whole genome resequencing of B. subtilis subti...        None        KEIO   \n",
       "2  Comprehensive identification and characterizat...        None      UT-MGS   \n",
       "3  Comprehensive identification and characterizat...        None      UT-MGS   \n",
       "4  Comprehensive identification and characterizat...        None      UT-MGS   \n",
       "\n",
       "                          center_project_name  \\\n",
       "0      Bacillus subtilis subsp. natto BEST195   \n",
       "1  Bacillus subtilis subsp. subtilis str. 168   \n",
       "2          Integrative Transcriptome Analysis   \n",
       "3          Integrateve Transcriptome Analysis   \n",
       "4          Integrative Transcriptome Analysis   \n",
       "\n",
       "                                   study_description  ... primary_study  \\\n",
       "0                                               None  ...          None   \n",
       "1  Whole genome resequencing of B. subtilis subti...  ...          None   \n",
       "2  Although recent studies have revealed that the...  ...          None   \n",
       "3  Although recent studies have revealed that the...  ...          None   \n",
       "4  Although recent studies have revealed that the...  ...          None   \n",
       "\n",
       "  sra_link               study_url_link                             xref_link  \\\n",
       "0     None                         None  pubmed: 20398357 || pubmed: 25329997   \n",
       "1     None                         None                      pubmed: 20398357   \n",
       "2     None  DBTSS: http://dbtss.hgc.jp/                      pubmed: 20400770   \n",
       "3     None  DBTSS: http://dbtss.hgc.jp/                      pubmed: 20400770   \n",
       "4     None  DBTSS: http://dbtss.hgc.jp/                      pubmed: 20400770   \n",
       "\n",
       "  study_entrez_link ddbj_link ena_link study_attribute submission_accession  \\\n",
       "0              None      None     None            None            DRA000001   \n",
       "1              None      None     None            None            DRA000002   \n",
       "2              None      None     None            None            DRA000003   \n",
       "3              None      None     None            None            DRA000004   \n",
       "4              None      None     None            None            DRA000005   \n",
       "\n",
       "         sradb_updated  \n",
       "0  2016-11-20 16:42:15  \n",
       "1  2017-09-21 20:23:35  \n",
       "2  2016-11-20 16:42:16  \n",
       "3  2016-11-20 16:42:16  \n",
       "4  2016-11-20 16:42:16  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_sql_query(\"SELECT * FROM study\", con)\n",
    "print(df.columns)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['experiment_ID', 'bamFile', 'fastqFTP', 'experiment_alias',\n",
      "       'experiment_accession', 'broker_name', 'center_name', 'title',\n",
      "       'study_name', 'study_accession', 'design_description', 'sample_name',\n",
      "       'sample_accession', 'sample_member', 'library_name', 'library_strategy',\n",
      "       'library_source', 'library_selection', 'library_layout',\n",
      "       'targeted_loci', 'library_construction_protocol', 'spot_length',\n",
      "       'adapter_spec', 'read_spec', 'platform', 'instrument_model',\n",
      "       'platform_parameters', 'sequence_space', 'base_caller',\n",
      "       'quality_scorer', 'number_of_levels', 'multiplier', 'qtype', 'sra_link',\n",
      "       'experiment_url_link', 'xref_link', 'experiment_entrez_link',\n",
      "       'ddbj_link', 'ena_link', 'experiment_attribute', 'submission_accession',\n",
      "       'sradb_updated'],\n",
      "      dtype='object')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>experiment_ID</th>\n",
       "      <th>bamFile</th>\n",
       "      <th>fastqFTP</th>\n",
       "      <th>experiment_alias</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>broker_name</th>\n",
       "      <th>center_name</th>\n",
       "      <th>title</th>\n",
       "      <th>study_name</th>\n",
       "      <th>study_accession</th>\n",
       "      <th>...</th>\n",
       "      <th>qtype</th>\n",
       "      <th>sra_link</th>\n",
       "      <th>experiment_url_link</th>\n",
       "      <th>xref_link</th>\n",
       "      <th>experiment_entrez_link</th>\n",
       "      <th>ddbj_link</th>\n",
       "      <th>ena_link</th>\n",
       "      <th>experiment_attribute</th>\n",
       "      <th>submission_accession</th>\n",
       "      <th>sradb_updated</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>DRX000001</td>\n",
       "      <td>DRX000001</td>\n",
       "      <td>None</td>\n",
       "      <td>KEIO</td>\n",
       "      <td>B. subtilis subsp. natto genome sequencing Sep...</td>\n",
       "      <td>DRP000001</td>\n",
       "      <td>DRP000001</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>DRA000001</td>\n",
       "      <td>2016-11-19 08:00:55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2.0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>DRX000002</td>\n",
       "      <td>DRX000002</td>\n",
       "      <td>None</td>\n",
       "      <td>KEIO</td>\n",
       "      <td>B. subtilis subsp. subtilis genome resequencin...</td>\n",
       "      <td>DRP000002</td>\n",
       "      <td>DRP000002</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>DRA000002</td>\n",
       "      <td>2016-11-19 08:00:55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3.0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>DRX000003</td>\n",
       "      <td>DRX000003</td>\n",
       "      <td>None</td>\n",
       "      <td>UT-MGS</td>\n",
       "      <td>DLD1_normoxia_nucleosome</td>\n",
       "      <td>DRP000003</td>\n",
       "      <td>DRP000003</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>DRA000003</td>\n",
       "      <td>2016-11-19 08:00:55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4.0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>DRX000004</td>\n",
       "      <td>DRX000004</td>\n",
       "      <td>None</td>\n",
       "      <td>UT-MGS</td>\n",
       "      <td>DLD1_polysome</td>\n",
       "      <td>DRP000004</td>\n",
       "      <td>DRP000004</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>DRA000004</td>\n",
       "      <td>2016-11-19 08:00:55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5.0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>DRX000005</td>\n",
       "      <td>DRX000005</td>\n",
       "      <td>None</td>\n",
       "      <td>UT-MGS</td>\n",
       "      <td>DLD1_cytoplasmic</td>\n",
       "      <td>DRP000005</td>\n",
       "      <td>DRP000005</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>DRA000005</td>\n",
       "      <td>2016-11-19 08:00:55</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   experiment_ID bamFile fastqFTP experiment_alias experiment_accession  \\\n",
       "0            1.0    None     None        DRX000001            DRX000001   \n",
       "1            2.0    None     None        DRX000002            DRX000002   \n",
       "2            3.0    None     None        DRX000003            DRX000003   \n",
       "3            4.0    None     None        DRX000004            DRX000004   \n",
       "4            5.0    None     None        DRX000005            DRX000005   \n",
       "\n",
       "  broker_name center_name                                              title  \\\n",
       "0        None        KEIO  B. subtilis subsp. natto genome sequencing Sep...   \n",
       "1        None        KEIO  B. subtilis subsp. subtilis genome resequencin...   \n",
       "2        None      UT-MGS                           DLD1_normoxia_nucleosome   \n",
       "3        None      UT-MGS                                      DLD1_polysome   \n",
       "4        None      UT-MGS                                   DLD1_cytoplasmic   \n",
       "\n",
       "  study_name study_accession  ... qtype sra_link experiment_url_link  \\\n",
       "0  DRP000001       DRP000001  ...  None     None                None   \n",
       "1  DRP000002       DRP000002  ...  None     None                None   \n",
       "2  DRP000003       DRP000003  ...  None     None                None   \n",
       "3  DRP000004       DRP000004  ...  None     None                None   \n",
       "4  DRP000005       DRP000005  ...  None     None                None   \n",
       "\n",
       "  xref_link experiment_entrez_link ddbj_link ena_link experiment_attribute  \\\n",
       "0      None                   None      None     None                 None   \n",
       "1      None                   None      None     None                 None   \n",
       "2      None                   None      None     None                 None   \n",
       "3      None                   None      None     None                 None   \n",
       "4      None                   None      None     None                 None   \n",
       "\n",
       "  submission_accession        sradb_updated  \n",
       "0            DRA000001  2016-11-19 08:00:55  \n",
       "1            DRA000002  2016-11-19 08:00:55  \n",
       "2            DRA000003  2016-11-19 08:00:55  \n",
       "3            DRA000004  2016-11-19 08:00:55  \n",
       "4            DRA000005  2016-11-19 08:00:55  \n",
       "\n",
       "[5 rows x 42 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_sql_query(\"SELECT * FROM experiment\", con)\n",
    "print(df.columns)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Other                       110102\n",
       "Whole Genome Sequencing      70056\n",
       "Metagenomics                 33643\n",
       "Transcriptome Analysis       32561\n",
       "Population Genomics            796\n",
       "Epigenetics                    620\n",
       "Exome Sequencing               281\n",
       "Cancer Genomics                181\n",
       "Pooled Clone Sequencing         31\n",
       "Synthetic Genomics              10\n",
       "Transcriptome Sequencing         2\n",
       "Whole Genome Sequencing          1\n",
       "Name: study_type, dtype: int64"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.study_type.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>100</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8199295</th>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8199296</th>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8199297</th>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8199298</th>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8199299</th>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         100\n",
       "8199295  100\n",
       "8199296  100\n",
       "8199297  100\n",
       "8199298  100\n",
       "8199299  100"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df[df.study_type == \"Transcriptome Analysis\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<sqlite3.Cursor at 0x2ad27ab3b6c0>"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# table = \"sample\"\n",
    "# cols = \"scientific_name,sample_accession\"\n",
    "# cursor.execute(f\"create virtual table {table}_ft using fts5({cols})\")\n",
    "# cursor.execute(f\"insert into {table}_ft select {cols} from {table}\")\n",
    "\n",
    "# table = \"experiment\"\n",
    "# cols = \"title,design_description,sample_accession,library_name,library_strategy\"\n",
    "# cursor.execute(f\"create virtual table {table}_ft using fts5({cols})\")\n",
    "# cursor.execute(f\"insert into {table}_ft select {cols} from {table}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "cols = \"title,design_description,sample_accession,library_name,library_strategy\"\n",
    "exp_map = {i:ix for ix, i in enumerate(cols.split(\",\"))}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Query Process"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['SRS4102342']\n"
     ]
    }
   ],
   "source": [
    "# def search_genus_name(genus):\n",
    "\n",
    "genus=\"Cronobacter\"\n",
    "\n",
    "query = f\"\"\"SELECT sample_accession \n",
    "        FROM sample_ft\n",
    "        WHERE scientific_name\n",
    "        MATCH '{genus}'\n",
    "        \"\"\"\n",
    "samples = \",\".join([\"'\" + i[0] + \"'\" for i in cursor.execute(query).fetchall()])\n",
    "\n",
    "query = f\"\"\"SELECT *\n",
    "        FROM experiment_ft\n",
    "        WHERE sample_accession IN ({samples})\n",
    "        AND experiment_ft MATCH 'RNA AND seq'\n",
    "        \"\"\"\n",
    "\n",
    "# print(query)\n",
    "\n",
    "results = cursor.execute(query).fetchall()\n",
    "\n",
    "positive_terms = \"wild WT\".split()\n",
    "negative_terms = \"plasmid\".split()\n",
    "\n",
    "new_results = []\n",
    "for i in results:\n",
    "    descriptive_text = i[exp_map[\"title\"]] + i[exp_map[\"library_name\"]]\n",
    "    add = True\n",
    "    for n in negative_terms:\n",
    "        if n in descriptive_text:\n",
    "            add = False\n",
    "            break\n",
    "    if add:\n",
    "        new_results.append(i)\n",
    "\n",
    "results = new_results\n",
    "\n",
    "samples = []\n",
    "\n",
    "for ix, i in enumerate(results):\n",
    "    descriptive_text = i[exp_map[\"title\"]] + i[exp_map[\"library_name\"]]\n",
    "    for p in positive_terms:\n",
    "        if p in descriptive_text:\n",
    "            samples.append(i[exp_map[\"sample_accession\"]])\n",
    "print(samples)\n",
    "\n",
    "def get_wigs_from_ftps(ftps, path=\".\"):\n",
    "    path = Path(path)\n",
    "    for file in ftps:\n",
    "        extracted = False\n",
    "        wget.download(file)\n",
    "        file = file.split(\"/\")[-1]\n",
    "        with tarfile.open(file, \"r:\") as tar:\n",
    "            names = tar.getmembers()\n",
    "            wig_files = [n for n in names if \".wig\" in n.name]\n",
    "            if len(wig_files) > 0:\n",
    "                tar.extractall(path=path, members=wig_files)\n",
    "                \n",
    "        os.remove(file)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "r = cursor.execute(f\"select distinct library_layout from {table}\")\n",
    "rs = list(r.fetchall())\n",
    "rs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'PAIRED - NOMINAL_SDEV: 24.7558; NOMINAL_LENGTH: 163; '"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rs[0][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# [[i for i in r[0].split(\";\") if \"NOMINAL\" not in i]\n",
    "max([len(list(r)) for r in rs])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "experiment_ID: 1225035.0\n",
      "bamFile: None\n",
      "fastqFTP: None\n",
      "experiment_alias: GSM1104375_1\n",
      "experiment_accession: SRX254733\n",
      "broker_name: None\n",
      "center_name: GEO\n",
      "title: GSM1104375: Galaxy1-[CV104_pHDB3_5.txt]; Escherichia coli str. K-12 substr. MG1655; RNA-Seq\n",
      "study_name: GSE45443\n",
      "study_accession: SRP019956\n",
      "design_description: None\n",
      "sample_name: GSM1104375\n",
      "sample_accession: SRS403764\n",
      "sample_member: None\n",
      "library_name: None\n",
      "library_strategy: RNA-Seq\n",
      "library_source: TRANSCRIPTOMIC\n",
      "library_selection: cDNA\n",
      "library_layout: SINGLE - \n",
      "targeted_loci: None\n",
      "library_construction_protocol: None\n",
      "spot_length: None\n",
      "adapter_spec: None\n",
      "read_spec: None\n",
      "platform: ILLUMINA\n",
      "instrument_model: Illumina HiSeq 2000\n",
      "platform_parameters: INSTRUMENT_MODEL: Illumina HiSeq 2000\n",
      "sequence_space: None\n",
      "base_caller: None\n",
      "quality_scorer: None\n",
      "number_of_levels: None\n",
      "multiplier: None\n",
      "qtype: None\n",
      "sra_link: None\n",
      "experiment_url_link: GEO Sample: http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM1104375\n",
      "xref_link: gds: 301104375\n",
      "experiment_entrez_link: None\n",
      "ddbj_link: None\n",
      "ena_link: None\n",
      "experiment_attribute: GEO Accession: GSM1104375\n",
      "submission_accession: SRA071482\n",
      "sradb_updated: 2016-11-19 11:12:01\n"
     ]
    }
   ],
   "source": [
    "table = \"experiment\"\n",
    "field = \"study_accession\" #\"sample_accession\"\n",
    "query = \"SRP019956\"\n",
    "\n",
    "cs = [i[0] for i in cursor.execute(f\"select * from {table}\").description]\n",
    "r = cursor.execute(f\"select * from {table} where {field} = '{query}'\")\n",
    "rs = list(r.fetchone())\n",
    "for ix , i in enumerate(cs):\n",
    "    if True: # i in \"library_strategy title design_description sample_accession\".split():\n",
    "        print(f\"{i}: {rs[ix]}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "may want to use the sample scientific name to boost the search of the experiment\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Scratch Work"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for ix , i in enumerate(cs):\n",
    "    if True: # i in \"library_strategy title design_description sample_accession\".split():\n",
    "        print(f\"{i}: {rs[ix]}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def has_rna(row):\n",
    "#     print(row)\n",
    "    if row is None:\n",
    "        return False\n",
    "    if \"RNA\" in row or \"rna\" in row:\n",
    "        return True\n",
    "    else:\n",
    "        return False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['sra_ID', 'SRR_bamFile', 'SRX_bamFile', 'SRX_fastqFTP', 'run_ID',\n",
       "       'run_alias', 'run_accession', 'run_date', 'updated_date', 'spots',\n",
       "       'bases', 'run_center', 'experiment_name', 'run_url_link',\n",
       "       'run_entrez_link', 'run_attribute', 'experiment_ID', 'experiment_alias',\n",
       "       'experiment_accession', 'experiment_title', 'study_name', 'sample_name',\n",
       "       'design_description', 'library_name', 'library_strategy',\n",
       "       'library_source', 'library_selection', 'library_layout',\n",
       "       'library_construction_protocol', 'adapter_spec', 'read_spec',\n",
       "       'platform', 'instrument_model', 'instrument_name',\n",
       "       'platform_parameters', 'sequence_space', 'base_caller',\n",
       "       'quality_scorer', 'number_of_levels', 'multiplier', 'qtype',\n",
       "       'experiment_url_link', 'experiment_entrez_link', 'experiment_attribute',\n",
       "       'sample_ID', 'sample_alias', 'sample_accession', 'taxon_id',\n",
       "       'common_name', 'anonymized_name', 'individual_name', 'description',\n",
       "       'sample_url_link', 'sample_entrez_link', 'sample_attribute', 'study_ID',\n",
       "       'study_alias', 'study_accession', 'study_title', 'study_type',\n",
       "       'study_abstract', 'center_project_name', 'study_description',\n",
       "       'study_url_link', 'study_entrez_link', 'study_attribute',\n",
       "       'related_studies', 'primary_study', 'submission_ID',\n",
       "       'submission_accession', 'submission_comment', 'submission_center',\n",
       "       'submission_lab', 'submission_date', 'sradb_updated'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>run_url_link</th>\n",
       "      <th>run_entrez_link</th>\n",
       "      <th>experiment_ID</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>experiment_url_link</th>\n",
       "      <th>experiment_entrez_link</th>\n",
       "      <th>sample_ID</th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>sample_url_link</th>\n",
       "      <th>sample_entrez_link</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>1.0</td>\n",
       "      <td>DRX000001</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>1.0</td>\n",
       "      <td>DRS000001</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2.0</td>\n",
       "      <td>DRX000002</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2.0</td>\n",
       "      <td>DRS000002</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  run_url_link run_entrez_link  experiment_ID experiment_accession  \\\n",
       "0         None            None            1.0            DRX000001   \n",
       "1         None            None            2.0            DRX000002   \n",
       "\n",
       "  experiment_url_link experiment_entrez_link  sample_ID sample_accession  \\\n",
       "0                None                   None        1.0        DRS000001   \n",
       "1                None                   None        2.0        DRS000002   \n",
       "\n",
       "  sample_url_link sample_entrez_link  \n",
       "0            None               None  \n",
       "1            None               None  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table[\"run_url_link run_entrez_link experiment_ID experiment_accession experiment_url_link experiment_entrez_link sample_ID sample_accession sample_url_link sample_entrez_link\".split()].head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = table[table[\"study_type\"].apply(lambda row: has_rna(row))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['WGS', 'FL-cDNA', 'CLONE', 'OTHER', 'POOLCLONE', 'ChIP-Seq', 'EST',\n",
       "       'MNase-Seq', 'WCS', 'AMPLICON', 'RNA-Seq', 'CTS', 'WXS',\n",
       "       'Bisulfite-Seq', 'MeDIP-Seq', 'miRNA-Seq', 'WGA', None,\n",
       "       'FAIRE-seq', 'CLONEEND', 'ncRNA-Seq', 'RIP-Seq', 'RAD-Seq',\n",
       "       'DNase-Hypersensitivity', 'SELEX', 'MBD-Seq', 'Targeted-Capture',\n",
       "       'Tn-Seq', 'ATAC-seq', 'Synthetic-Long-Read',\n",
       "       'Tethered Chromatin Conformation Capture', 'FINISHING', 'Hi-C',\n",
       "       'ChIP', 'MRE-Seq', 'VALIDATION',\n",
       "       'Targeted resequencing of human colorectal cancer (CRC) stem-like cells (CCSCs), stem-like circulating tumor cells (cCCSCs) and metastatic colorectal stem cells (mCCSCs).',\n",
       "       'other', 'ChIA-PET'], dtype=object)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table.library_strategy.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>run_url_link</th>\n",
       "      <th>run_entrez_link</th>\n",
       "      <th>experiment_ID</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>experiment_url_link</th>\n",
       "      <th>experiment_entrez_link</th>\n",
       "      <th>sample_ID</th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>sample_url_link</th>\n",
       "      <th>sample_entrez_link</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [run_url_link, run_entrez_link, experiment_ID, experiment_accession, experiment_url_link, experiment_entrez_link, sample_ID, sample_accession, sample_url_link, sample_entrez_link]\n",
       "Index: []"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x[\"run_url_link run_entrez_link experiment_ID experiment_accession experiment_url_link experiment_entrez_link sample_ID sample_accession sample_url_link sample_entrez_link\".split()].head(2)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8",
   "language": "python",
   "name": "python-3.8"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
